app/src/main/java/com/marvel/marvel/budget/model/resolutionstrategy/GreedyStrategy.kt

```diff
@@ -0,0 +1,93 @@
+package com.marvel.marvel.budget.model.resolutionstrategy
+
+import com.marvel.marvel.viewmodel.ComicsViewModel
+import java.math.BigDecimal
+import java.util.*
+
+class GreedyStrategy : Strategy {
```

Comment:
This class looks really bad. Also unpacking everything from api as nullable and propagating this nullability is bad. This is by untested version of this class:
```
class GreedyStrategy : Strategy {

    override fun sortComicsByPricePerPage(comicsList: List<ComicsViewModel>, budgetAsString: String?): List<ComicsViewModel> {
        val budget = budgetAsString?.toBigDecimalOrNull() ?: BigDecimal.ZERO
        val pricesPerPage = comicsList
                .filter { it.pageCount != 0 }
                .mapNotNull(this::calculatePricePerPage)
                .sortedBy { it.pricePerPage }
        return takeFirstComicsInBudget(pricesPerPage, budget)
    }

    private fun takeFirstComicsInBudget(pricePerPages: List<PricePerPage>, budget: BigDecimal): List<ComicsViewModel> {
        val (comic, price) = pricePerPages.first()
        return when {
            price <= budget -> listOf(comic) + takeFirstComicsInBudget(pricePerPages.drop(1), budget - price)
            else -> return emptyList()
        }
    }

    private fun calculatePricePerPage(model: ComicsViewModel): PricePerPage? {
        val price: BigDecimal = model.price?.toBigDecimalOrNull() ?: return null
        val pricePerPage: Double = calcualtePricePerPage(model.price, model.pageCount) ?: return null
        return PricePerPage(model, price, pricePerPage)
    }

    // I should probably check out what does API return, but on the end I just want to devide this two numbers or have null
    private fun calcualtePricePerPage(priceString: String?, pages: Int?): Double? = try {
        priceString!!.toDouble() / pages!!
    } catch (_: Throwable) {
        null
    }

    data class PricePerPage(val comics: ComicsViewModel, val price: BigDecimal, val pricePerPage: Double)
}
```

Reply:
Thanks a lot Marcin, this is full of precious suggestions!